runtime/master/src/main/java/org/apache/nemo/runtime/master/scheduler/SimulationScheduler.java

```diff
@@ -0,0 +1,579 @@
+      if (!tasksToSchedule.isEmpty()) {
+        LOG.info("Scheduling some tasks in {}, which are in the same ScheduleGroup", tasksToSchedule.stream()
+          .map(Task::getTaskId)
+          .map(RuntimeIdManager::getStageIdFromTaskId)
+          .collect(Collectors.toSet()));
+
+        // Set the pointer to the schedulable tasks.
+        pendingTaskCollectionPointer.setToOverwrite(tasksToSchedule);
+
+        // Notify the dispatcher that a new collection is available.
+        taskDispatcher.onNewPendingTaskCollectionAvailable();
+      }
+    } else {
+      LOG.info("Skipping this round as no ScheduleGroup is schedulable.");
+    }
+  }
+
+  @Override
+  public void updatePlan(final PhysicalPlan newPhysicalPlan) {
+    // update the physical plan in the scheduler.
+    // NOTE: what's already been executed is not modified in the new physical plan.
+    // TODO #182: Consider reshaping in run-time optimization. At now, we only consider plan appending.
+    updatePlan(newPhysicalPlan, planStateManager.getMaxScheduleAttempt());
+  }
+
+  /**
+   * Update the physical plan in the scheduler.
+   *
+   * @param newPhysicalPlan    the new physical plan to update.
+   * @param maxScheduleAttempt the maximum number of task scheduling attempt.
+   */
+  private void updatePlan(final PhysicalPlan newPhysicalPlan,
+                          final int maxScheduleAttempt) {
+    planStateManager.updatePlan(newPhysicalPlan, maxScheduleAttempt);
+    this.sortedScheduleGroups = newPhysicalPlan.getStageDAG().getVertices().stream()
+      .collect(Collectors.groupingBy(Stage::getScheduleGroup))
+      .entrySet().stream()
+      .sorted(Map.Entry.comparingByKey())
+      .map(Map.Entry::getValue)
+      .collect(Collectors.toList());
+  }
+
+  @Override
+  public void onExecutorAdded(final ExecutorRepresenter executorRepresenter) {
+    LOG.info("{} added (node: {})", executorRepresenter.getExecutorId(), executorRepresenter.getNodeName());
+    executorRegistry.registerExecutor(executorRepresenter);
+    this.simulatedTaskExecutorMap.put(executorRepresenter.getExecutorId(),
+      new SimulatedTaskExecutor(this, executorRepresenter, actualMetricStore));
+    taskDispatcher.onExecutorSlotAvailable();
+  }
+
+  @Override
+  public void onExecutorRemoved(final String executorId) {
+    // Role of FailedEvaluatorHandler + onExecutorFailed.
+    metricCountDownLatch.countDown();
+    LOG.info("{} removed", executorId);
+
+    // These are tasks that were running at the time of executor removal.
+    final Set<String> interruptedTasks = new HashSet<>();
+    executorRegistry.updateExecutor(executorId, (executor, state) -> {
+      interruptedTasks.addAll(executor.onExecutorFailed());
+      return Pair.of(executor, ExecutorRegistry.ExecutorState.FAILED);
+    });
+
+    // Retry the interrupted tasks (and required parents)
+    SchedulerUtils.retryTasksAndRequiredParents(planStateManager, blockManagerMaster, interruptedTasks);
+
```

Comment:
I guess if something is added to the place in the batch scheduler corresponding to this line, then the same code should be added to this line?

Reply:
I tried to refactor the common parts to the BatchSchedulerUtil, so it seems like this would be alright.